### ISSUE TITLE
What is 6th unary operator?

### ISSUE THREAD

**Issue by khiri-co** [rust-lang/rust#20876]
In reference 7.2.11 Unary operator expressions
> Rust defines six symbolic unary operators.

But in document there is only 5 ops; '-', '*', '!', 'box', '&'.
What is the 6th unary op?

By the way, in Rust '+' is not an unary op. So this expression is invalid; 'let a = +(1+2);'.
Considering the symmetry of "+" & '-', I think '+' may be an unary operator.
Of course, '+' unary op has no effect.

**Comment by nagisa** [comment]
`..` is unary operator in some contexts.

**Comment by japaric** [comment]
TL;DR Probably the deprecated `~` operator, that was used to box things. `~"foo"` created a `Box<str>`, where as `~[1, 2]` created a `Box<[T]>`.

According to the compiler internals, we have [4 unary operators](http://doc.rust-lang.org/syntax/ast/enum.UnOp.html) (in the AST):

(Operator name : [expression](http://doc.rust-lang.org/syntax/ast/enum.Expr_.html#variant.ExprUnary))
- `UnUniq`: `~"string"` (which is deprecated syntax)
- `UnDeref`: `*boxed_thing`
- `UnNot`: `!bits`
- `UnNeg`: `-num`

From the POV of the compiler, the feature gated `box thing` is an [expression](http://doc.rust-lang.org/syntax/ast/enum.Expr_.html#variant.ExprBox). And references `&foo`/`&mut bar` are also [expressions](http://doc.rust-lang.org/syntax/ast/enum.Expr_.html#variant.ExprAddrOf).

You may think of `&` as the "address of" operator, and `box` as the "box" operator. But that's not quite right, because when you use them in patterns like `let &e = &(2, 3)`, they don't perform any operation at all.

**Comment by japaric** [comment]
(Oh and the reference is probably out of date in several places, we haven't put effort in updating it yet)
